FFI functions for Rust bindings

niklasdewally opened the pull request:
@ChrisJefferson 
The corresponding C++ for https://github.com/conjure-cp/conjure-oxide/pull/28.

Mainly wraps things as FFI safe functions - i.e. using pointers for complex types, re-exporting inline functions, manual constructors and destructors, ....

niklasdewally committed:
Add minion.h to libwrapper.h

niklasdewally committed:
Export constructors for SearchOptions and SearchMethods in libwrapper.h

niklasdewally committed:
Export CSPInstance constructor in libwrapper.h

niklasdewally committed:
Add C-style FFI wrapper functions in libminion.h

niklasdewally committed:
Enforce single-threaded library use through a mutex

niklasdewally committed:
Edit documentation in libwrapper.h

niklasdewally commented:
(this PR should probably be squash merged)

ChrisJefferson commented:
This seems very reasonable. I'll just wait for the tests to spin, them I'm happy to merge. There are some tweaks that can be done, but I think getting to a MVP, then tuning, is a good idea.

ChrisJefferson commented on minion/libwrapper.cpp:
```diff
@@ -194,6 +197,109 @@ void finaliseModel(CSPInstance& instance) {
+
+int printMatrix_getValue(int idx) {
+  // FIXME: does not work with minion debug mode
+  return globals->state_m->getPrintMatrix()[idx][0].assignedValue();
```
Replace this with:  return checked_cast<int>(globals->state_m->getPrintMatrix()[idx][0].assignedValue());

checked_cast is a template I wrote (it's designed to look like a C++ cast, if you are familiar with them).

In debug mode, DomainInt does overflow checking (which C++ doesn't do for normal ints), and you need to explictly turn it back into an 'int'.

niklasdewally committed:
Use checked_cast in libminion.cpp to fix debug build

niklasdewally commented:
@ChrisJefferson I've added the checked_cast as discussed - would appreciate another review.

I've fixed git history since last review, so should no longer need a squash merge.
